core/server/master/src/main/java/alluxio/underfs/UfsStatusCache.java

```diff
@@ -0,0 +1,275 @@
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Future;
+import java.util.concurrent.RejectedExecutionException;
+
+/**
+ * This class is a cache from an Alluxio namespace URI ({@link AlluxioURI}, i.e. /path/to/inode) to
+ * UFS statuses.
+ *
+ * It also allows associating a path with child inodes, so that the statuses for a specific path can
+ * be searched for later.
+ */
+public class UfsStatusCache {
+  private static final Logger LOG = LoggerFactory.getLogger(UfsStatusCache.class);
+
+  private final ConcurrentHashMap<AlluxioURI, UfsStatus> mStatuses;
+  private final ConcurrentHashMap<AlluxioURI, Future<Collection<UfsStatus>>> mActivePrefetchJobs;
+  private final ConcurrentHashMap<AlluxioURI, Collection<UfsStatus>> mChildren;
+  private final ExecutorService mPrefetchExecutor;
+
+  /**
+   * Create a new instance of {@link UfsStatusCache}.
+   *
+   * @param prefetchExecutor the executor service used to prefetch statuses
+   */
+  public UfsStatusCache(@Nullable ExecutorService prefetchExecutor) {
+    mStatuses = new ConcurrentHashMap<>();
+    mChildren = new ConcurrentHashMap<>();
+    mActivePrefetchJobs = new ConcurrentHashMap<>();
+    mPrefetchExecutor = prefetchExecutor;
+  }
+
+  /**
+   * Add a new status to the cache.
+   *
+   * The last component of the path in the {@link AlluxioURI} must match the result of
+   * {@link UfsStatus#getName()}. This method overrides any status currently cached for the same
+   * URI.
+   *
+   * @param path the Alluxio path to key on
+   * @param status the ufs status to store
+   * @throws IllegalArgumentException if the status name doesn't match the final URI path component
+   */
+  public void addStatus(AlluxioURI path, UfsStatus status) {
+    UfsStatus prev = mStatuses.putIfAbsent(path, status);
+    if (!path.getName().equals(status.getName())) {
+      throw new IllegalArgumentException(
+          String.format("path name %s does not match ufs status name %s",
+              path.getName(), status.getName()));
+    }
+  }
+
+  /**
+   * Add a parent-child mapping to the status cache.
+   *
+   * All child statuses added via this method will be available via {@link #getStatus(AlluxioURI)}.
+   *
+   * @param path the directory inode path which contains the children
+   * @param children the children of the {@code path}
+   * @throws IllegalArgumentException when {@code path} already exists or if any child already
+   *                                  exists
+   */
+  public void addChildren(AlluxioURI path, Collection<UfsStatus> children) {
+    UfsStatus status = mStatuses.get(path);
+    // If this path doesn't yet exist, we can't keep track of the parent-child relationship
+    // We can still add statuses to the cache regardless
+    if (status != null) {
```

Comment:
Why can't we add this parent->children listing cache even if the parent status is not in the other map?

Reply:
ah, this was a limitation of my previous implementation. It is possible now.

Reply:
updated